# Home Assignment 5

## Problem 1 (3 points). Shamir secret sharing

Find the secret and initial polynomial in (n, 4)-threshold Shamir secret sharing scheme over $F_{19}$ for $(3,17); (17,5); (5,11); (11,4)$. Is it consistent with $(4, 3)$?

**Given**:

$F(3) = 17$, $F(5) = 11$, $F(11) = 4$, $F(17) = 5$, $p = 19$

$Secret = F(0) = ?$

**Solution**:

$F(x) = \sum_{i=1}^{4} y_{i} \times l_{x_{i}}(x)$

$l_{x_{i}} = \prod_{i \neq j} \frac{x - x_{j}}{x_{i} - x_{j}}$

$F(x) = 17\frac {(x - 5)(x - 11)(x - 17)} {(3 - 5)(3 - 11)(3 - 17)} + 11\frac {(x - 3)(x - 11)(x - 17)} {(5 - 3)(5 - 11)(5 - 17)} + 4\frac {(x - 3)(x - 5)(x - 17)} {(11 - 3)(11 - 5) (11 - 17)} + 5\frac {(x - 3)(x - 5)(x - 11)}{(17 - 3)(17 - 5)(17 - 11)} = 17 \frac {(x - 5)(x - 11)(x - 17)} {-224} + 11 \frac {(x - 3)(x - 11)(x - 17)} {144} + 4 \frac {(x - 3)(x - 5)(x - 17)} {-288} + 5 \frac {(x - 3)(x - 5)(x - 11)} {1008} = - 17*(x - 5)(x - 11)(x - 17)*224^{-1} + 11*(x - 3)(x - 11)(x - 17)*144^{-1} - 4*(x - 3)(x - 5)(x - 17)*288^{-1} + 5*(x - 3)(x - 5)(x - 11)*1008^{-1} = 9(x - 5)(x - 11)(x - 17) + (x - 3)(x - 11)(x - 17) + 5(x - 3)(x - 5)(x - 17) + 5(x - 3)(x - 5)(x - 11) = 20x^3 - 548x^2 + 4484x - 11076 = x^3 + 3x^2 + 1$

**Initial polynomial**:
$x^3 +3x^2 + 1$

**Secret**: $F(0) = 1$

$F(4) = 4^3 + 3*4^2 + 1 = 7 + 10 + 1 = 18$

$18 \neq 3 \therefore$ polynomial is not consistent with $(4,3)$

## Problem 2 (2* points). Schnorr ring signature

For the elliptic curve $y^2=x^3+7$ over $F_{127}$ and the hash function $H() = SHA256()\%127$. Consider a subgroup of point on a curve with a generator $G = (19, 32)$.

Suppose Alice has a private key $a = 17$ and Bob has a private key $b = 29$.

1. (0 points) Compute public keys.
2. (2* points) Generate Schnorr's ring signature for $m=77$ on Alice's behalf with $k=101$.

1. Public key: P = aG

$P_{A} = 17 * (19, 32)$
$P_{B} = 29 * (19, 32)$

## Problem 3 (3 points). Bitcoin data

Consider tx=21e8fd77b81b6c4cf6723c10ea83acae77728f188ee6f4cd97ae114665b5fc0f (block height: 478994, tx index/position: 1603). Trace the transaction back to address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw.

Use blockexplorer Python library, Bitcoin RPC or their equivalent alternative.

In [26]:
import requests, json

tx = "21e8fd77b81b6c4cf6723c10ea83acae77728f188ee6f4cd97ae114665b5fc0f"
base_url = "https://blockchain.info/"
tx = requests.get(base_url + "rawtx/" + tx)
tx = json.loads(tx.text)

txs_step_1 = []
txs_step_2 = []
txs_step_3 = []

def iterate(txs_step_i, step):
    for input_addr in tx["inputs"]:
        prev_tx = requests.get(base_url + "rawtx/" + str(input_addr["prev_out"]["tx_index"]))
        prev_tx = json.loads(prev_tx.text)
        txs_step_i.append(prev_tx)
        for output in prev_tx["out"]:
            if output["addr"] == "12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw":
                if (step == 1 | step == 2):
                    print(f"Transaction {tx.tx_index} spent {output.value/1e8} BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw")
                    break
                if step == 3:
                    print("Found transaction with hash " + str(prev_tx["hash"]) + " that spends " + str(output["value"]/1e8) + " BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw")  
                    break

iterate(txs_step_1, 1)
            
for tx in txs_step_1:
    iterate(txs_step_2, 2)

for tx in txs_step_2:
    iterate(txs_step_3, 3)

Found transaction with hash 00a612aa7b2e35de0be7025b77c3934a4428fe951b34236701fdd48466f62992 that spends 0.17124673 BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw
Found transaction with hash 02b43152862ee98f00954eaf2cad6ecf8d6fe09dffb9f35be445dd8b10448f34 that spends 0.1801023 BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw
Found transaction with hash 04cc3f8cc717d27425359a61dbd527a4d29f34c1b13d4d189bfc00e6661ae808 that spends 0.16656716 BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw
Found transaction with hash 055a69935824db79df6fd42496c6a35f5f291add5196b306bf0a927b334584d8 that spends 0.1703 BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw
Found transaction with hash 0955b755d72821c9b1cfc9e7673710806244c56ab955c6c28c5630013846563b that spends 0.169 BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw
Found transaction with hash 0a9d3c83d6d380616bac17c687ed8e5a2bd48d761264753fae3fecf8bb264a03 that spends 0.1804 BTC from address 12t9YDPgwueZ9NyMgw519p7AA8isjr6SMw
Found 

## Problem 4 (3* points). Shared send untangling

1. (1* point) Simplify transaction (see Section 1 \[1\])
* 409803bb5e124fd028c0482027c7722e84ce55b78204b279d3a44aba5e7c1698
* 9ea6fdad2ea5870f5b17cb28587263f90cd7ec5c67f35d395a318186e5c79333
* c38aac9910f327700e0f199972eed8ea7c6b1920e965f9cb48a92973e7325046
* 659135664894e50040830edb516a76f704fd2be409ecd8d1ea9916c002ab28a2
2. (2* points) What are these transactions types (see Theorem 1)?

\[1\] https://bitfury.com/content/downloads/bitfury_whitepaper_shared_send_untangling_in_bitcoin_8_24_2016.pdf

## Problem 5 (4 points). ERC20 data

Consider (scam?) token 0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0 from Ethereum mainnet. How many tokens where transfered in Dec-2020?

In [7]:
import pandas as pd
data = pd.read_csv('Scam token dec-2020.csv')
data = data[data.Method == "Transfer"]
data

Txhash   Blockno  \
4     0x4cefffbfea31bda18eb20c77ad622e18333c50ed3084...  11461644   
5     0x3bd89b7fd0acf9df3743b8bfe12c75f51173265d1e90...  11461677   
6     0x5b1909ab79b6513aefcef1bd530e74d30be5bf6ddfd2...  11461710   
7     0xffe4322c3bdc48efa2b3332df04e31cb39d17200d957...  11461718   
8     0x7735c24a0a7584cb94a679de4cb8ef864c777a955867...  11461758   
...                                                 ...       ...   
2261  0x7a18e80993465fa9898c7cded940ee468aa63a9f6ec9...  11556561   
2266  0x1cf778fb2778b3ab9bc8a160e4518a93c73da9256a4e...  11558996   
2267  0xab9daab3a2e67b006abf480c5649a58f4dc06295e478...  11559000   
2272  0xfba9c1dd6042654388f65e388af24bb2ab05af00b663...  11561630   
2275  0xa8e7056a26899c372387f356788fe33eabbe88daabd2...  11563649   

      UnixTimestamp             DateTime  \
4        1608088590  2020-12-16 03:16:30   
5        1608089180  2020-12-16 03:26:20   
6        1608089519  2020-12-16 03:31:59   
7        1608089696  2020-12-16 03:34:56   
8        1608090107  2020-12-16 03:41:47   
...             ...                  ...   
2261     1609346669  2020-12-30 16:44:29   
2266     1609378817  2020-12-31 01:40:17   
2267     1609378868  2020-12-31 01:41:08   
2272     1609413763  2020-12-31 11:22:43   
2275     1609440685  2020-12-31 18:51:25   

                                            From  \
4     0x4775259e2cb64ad5bda0a396568732fc655c593c   
5     0xaf9981d21d212d9cb5f2c9df6b56029737ac53e7   
6     0xaf9981d21d212d9cb5f2c9df6b56029737ac53e7   
7     0x1bfeaf74553325fc7064f6d19287cb9374ce3f05   
8     0x3e9afaa4a062a49d64b8ab057b3cb51892e17ecb   
...                                          ...   
2261  0x6f3730a184054b96786dda5387ea70a9967732d1   
2266  0x4e0e8e636d01216855946aaa163d7fb73c86d070   
2267  0x3e9afaa4a062a49d64b8ab057b3cb51892e17ecb   
2272  0x141fef8cd8397a390afe94846c8bd6f4ab981c48   
2275  0x6f3730a184054b96786dda5387ea70a9967732d1   

                                              To  ContractAddress  \
4     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
5     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
6     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
7     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
8     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
...                                          ...              ...   
2261  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2266  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2267  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2272  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2275  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   

      Value_IN(ETH)  Value_OUT(ETH)  CurrentValue @ $1689.35/Eth  TxnFee(ETH)  \
4               0.0               0                          0.0     0.003797   
5               0.0               0                          0.0     0.003873   
6               0.0               0                          0.0     0.001699   
7               0.0               0                          0.0     0.003450   
8               0.0               0                          0.0     0.004413   
...             ...             ...                          ...          ...   
2261            0.0               0                          0.0     0.002519   
2266            0.0               0                          0.0     0.002154   
2267            0.0               0                          0.0     0.005184   
2272            0.0               0                          0.0     0.003894   
2275            0.0               0                          0.0     0.001217   

      TxnFee(USD)  Historical $Price/Eth Status ErrCode    Method  Unnamed: 16  
4        6.414611                 637.74    NaN     NaN  Transfer          NaN  
5        6.543106                 637.74    NaN     NaN  Transfer          NaN  
6        2.869648    

In [8]:
data.ErrCode.unique()

array([nan, 'Out of gas', 'Reverted'], dtype=object)

In [9]:
data.Status.unique()

array([nan, 'Error(0)'], dtype=object)

In [14]:
err_data = data[(data.ErrCode == 'Out of gas') | (data.ErrCode == 'Reverted')]
err_data

Txhash   Blockno  \
917   0xdb9dd43516fb9565bd7e0400f418253cf4b49474e3ea...  11477432   
1979  0x2ca0fbb1c544f722f3e09dfe2c90752e183bb41e8282...  11498506   
1991  0x3150184f22143156e6a0eda6aa689a0a9dbf2d96e131...  11498834   
2004  0x95f703c3643c1b0683d9829ba5d49dd326426c525106...  11499084   
2075  0xc1294519a0fb8a5332baf23248c91bdeb696ff2b997f...  11504597   

      UnixTimestamp             DateTime  \
917      1608297605  2020-12-18 13:20:05   
1979     1608577233  2020-12-21 19:00:33   
1991     1608581589  2020-12-21 20:13:09   
2004     1608584798  2020-12-21 21:06:38   
2075     1608657380  2020-12-22 17:16:20   

                                            From  \
917   0x36bc2a779283b0581728c1a52b1eeab4e2585b08   
1979  0x35821c98d3913377dbf90c1703d53d3ea17780ef   
1991  0x34ef2c22bd8047e72d8cb9f5517e96d0ee026dfd   
2004  0x54f62de0a9519fc9e98e3e807ac2ecb66d5d2061   
2075  0x6e3e80d88fd4e3417831b8686db83fc36dbe5fee   

                                              To  ContractAddress  \
917   0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
1979  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
1991  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2004  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2075  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   

      Value_IN(ETH)  Value_OUT(ETH)  CurrentValue @ $1689.35/Eth  TxnFee(ETH)  \
917        0.004159               0                     7.026007     0.000924   
1979       0.000000               0                     0.000000     0.000828   
1991       0.000000               0                     0.000000     0.003594   
2004       0.000000               0                     0.000000     0.000827   
2075       0.000000               0                     0.000000     0.002839   

      TxnFee(USD)  Historical $Price/Eth    Status     ErrCode    Method  \
917      1.560959                 654.55  Error(0)  Out of gas  Transfer   
1979     1.398241                 607.90  Error(0)    Reverted  Transfer   
1991     6.072362                 607.90  Error(0)    Reverted  Transfer   
2004     1.397532                 607.90  Error(0)    Reverted  Transfer   
2075     4.796403                 637.23  Error(0)    Reverted  Transfer   

      Unnamed: 16  
917           NaN  
1979          NaN  
1991          NaN  
2004          NaN  
2075          NaN

In [16]:
corrected_data = data[data.ErrCode.isna()]
corrected_data

Txhash   Blockno  \
4     0x4cefffbfea31bda18eb20c77ad622e18333c50ed3084...  11461644   
5     0x3bd89b7fd0acf9df3743b8bfe12c75f51173265d1e90...  11461677   
6     0x5b1909ab79b6513aefcef1bd530e74d30be5bf6ddfd2...  11461710   
7     0xffe4322c3bdc48efa2b3332df04e31cb39d17200d957...  11461718   
8     0x7735c24a0a7584cb94a679de4cb8ef864c777a955867...  11461758   
...                                                 ...       ...   
2261  0x7a18e80993465fa9898c7cded940ee468aa63a9f6ec9...  11556561   
2266  0x1cf778fb2778b3ab9bc8a160e4518a93c73da9256a4e...  11558996   
2267  0xab9daab3a2e67b006abf480c5649a58f4dc06295e478...  11559000   
2272  0xfba9c1dd6042654388f65e388af24bb2ab05af00b663...  11561630   
2275  0xa8e7056a26899c372387f356788fe33eabbe88daabd2...  11563649   

      UnixTimestamp             DateTime  \
4        1608088590  2020-12-16 03:16:30   
5        1608089180  2020-12-16 03:26:20   
6        1608089519  2020-12-16 03:31:59   
7        1608089696  2020-12-16 03:34:56   
8        1608090107  2020-12-16 03:41:47   
...             ...                  ...   
2261     1609346669  2020-12-30 16:44:29   
2266     1609378817  2020-12-31 01:40:17   
2267     1609378868  2020-12-31 01:41:08   
2272     1609413763  2020-12-31 11:22:43   
2275     1609440685  2020-12-31 18:51:25   

                                            From  \
4     0x4775259e2cb64ad5bda0a396568732fc655c593c   
5     0xaf9981d21d212d9cb5f2c9df6b56029737ac53e7   
6     0xaf9981d21d212d9cb5f2c9df6b56029737ac53e7   
7     0x1bfeaf74553325fc7064f6d19287cb9374ce3f05   
8     0x3e9afaa4a062a49d64b8ab057b3cb51892e17ecb   
...                                          ...   
2261  0x6f3730a184054b96786dda5387ea70a9967732d1   
2266  0x4e0e8e636d01216855946aaa163d7fb73c86d070   
2267  0x3e9afaa4a062a49d64b8ab057b3cb51892e17ecb   
2272  0x141fef8cd8397a390afe94846c8bd6f4ab981c48   
2275  0x6f3730a184054b96786dda5387ea70a9967732d1   

                                              To  ContractAddress  \
4     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
5     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
6     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
7     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
8     0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
...                                          ...              ...   
2261  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2266  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2267  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2272  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   
2275  0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0              NaN   

      Value_IN(ETH)  Value_OUT(ETH)  CurrentValue @ $1689.35/Eth  TxnFee(ETH)  \
4               0.0               0                          0.0     0.003797   
5               0.0               0                          0.0     0.003873   
6               0.0               0                          0.0     0.001699   
7               0.0               0                          0.0     0.003450   
8               0.0               0                          0.0     0.004413   
...             ...             ...                          ...          ...   
2261            0.0               0                          0.0     0.002519   
2266            0.0               0                          0.0     0.002154   
2267            0.0               0                          0.0     0.005184   
2272            0.0               0                          0.0     0.003894   
2275            0.0               0                          0.0     0.001217   

      TxnFee(USD)  Historical $Price/Eth Status ErrCode    Method  Unnamed: 16  
4        6.414611                 637.74    NaN     NaN  Transfer          NaN  
5        6.543106                 637.74    NaN     NaN  Transfer          NaN  
6        2.869648    

In [18]:
corrected_data.nunique()

Txhash                         608
Blockno                        501
UnixTimestamp                  501
DateTime                       501
From                           319
To                               1
ContractAddress                  0
Value_IN(ETH)                    1
Value_OUT(ETH)                   1
CurrentValue @ $1689.35/Eth      1
TxnFee(ETH)                    454
TxnFee(USD)                    454
Historical $Price/Eth           15
Status                           0
ErrCode                          0
Method                           1
Unnamed: 16                      0
dtype: int64

319 различных отправителей суммарно отправили 608 токенов для 0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0 в декабре 2020.

## Problem 6 (3* points). BigQuery

Consider (scam?) token 0xcae72a7a0fd9046cf6b165ca54c9e3a3872109e0 from Ethereum mainnet. Compute the total amount of tokens transfered. Use bigquery-public-data.crypto_ethereum (https://ethereum-etl.readthedocs.io/en/latest/google-bigquery/).